### PR TITLE
Some HTTP fixes factored out of the HTTP masterserver PR

### PR DESCRIPTION
**Pull request by heinrich5991**
CC #5064

### CODE REVIEW

**Comment by heinrich5991** [comment]
So that @Learath2 can refactor our HTTP stack to curl_multi to allow connection reuse.

From [CURLSHOPT_SHARE in the libcurl documentation](https://curl.se/libcurl/c/CURLSHOPT_SHARE.html) (emphasis mine):

>CURL_LOCK_DATA_CONNECT
>
>Put the connection cache in the share object and make all easy handles using this share object share the connection cache.
>
>Note that due to a known bug, it is not safe to share connections this way between multiple concurrent threads.
>
>**Connections that are used for HTTP/1.1 Pipelining or HTTP/2 multiplexing only get additional transfers added to them if the existing connection is held by the same multi or easy handle. libcurl does not support doing HTTP/2 streams in different threads using a shared connection.**
>
>Support for [CURL_LOCK_DATA_CONNECT](https://curl.se/libcurl/c/CURLSHOPT_SHARE.html#CURLLOCKDATACONNECT) was added in 7.57.0, but the symbol existed before this.
>
>Note that when you use the multi interface, all easy handles added to the same multi handle will share connection cache by default without using this option.

**Comment by edg-l** [comment]
Seems like you forgot to include some file to cmake

> ENGINE_SHARED does not contain every file from directory src/engine/shared

**Comment by Learath2** [comment]
Do you want me to review this on it's own merits?

**Comment by heinrich5991** [comment]
> Do you want me to review this on it's own merits?

Haven't thought about it, but yes.

File: src/engine/shared/http.cpp [hunk 1]
```diff
@@ -183,174 +232,111 @@ int CRequest::RunImpl(CURL *pHandle)
...
+size_t CHttpRequest::OnData(char *pData, size_t DataSize)
 {
-	return ((CRequest *)pUser)->OnData(pData, Size * Number);
+	if(!m_WriteToFile)
+	{
+		if(DataSize == 0)
+		{
+			return DataSize;
+		}
+		bool Reallocate = false;
```
Comment (Learath2) [inline src/engine/shared/http.cpp:243]: How about `size_t NewBufferSize = min(1024, m_BufferLength); while(m_BufferLength + DataSize > NewBufferSize) NewBufferSize *= 2; if(NewBufferSize != m_BufferSize) realloc;`?

**Review by Learath2 — APPROVED** [review]
LGTM except for the small comment I made, I would be okay with it as-is too.

**Comment by heinrich5991** [comment]
bors retry

**Pull request merged**